src/ray/raylet/local_object_manager.cc

```diff
@@ -33,15 +34,28 @@ void LocalObjectManager::PinObjects(const std::vector<ObjectID> &object_ids,
                      << " was evicted before the raylet could pin it.";
       continue;
     }
-    RAY_LOG(DEBUG) << "Pinning object " << object_id;
-    pinned_objects_size_ += object->GetSize();
-    pinned_objects_.emplace(object_id, std::make_pair(std::move(object), owner_address));
-  }
-}
 
-void LocalObjectManager::WaitForObjectFree(const rpc::Address &owner_address,
-                                           const std::vector<ObjectID> &object_ids) {
-  for (const auto &object_id : object_ids) {
+    const auto inserted = objects_waiting_for_free_.emplace(object_id, owner_address);
+    if (inserted.second) {
+      // This is the first time we're pinning this object.
+      RAY_LOG(DEBUG) << "Pinning object " << object_id;
+      pinned_objects_size_ += object->GetSize();
+      pinned_objects_.emplace(object_id, std::move(object));
+    } else {
+      if (inserted.first->second.worker_id() != owner_address.worker_id()) {
+        // TODO(swang): Handle this case. We should use the new owner address
+        // and object copy.
+        auto original_worker_id =
+            WorkerID::FromBinary(inserted.first->second.worker_id());
+        auto new_worker_id = WorkerID::FromBinary(owner_address.worker_id());
+        RAY_LOG(WARNING)
```

Comment:
Are we going to do this in a follow up? Also, is there any other case that owner address is needed from the raylet?

Reply:
I think a better way might be to use nondeterministic ObjectIDs so we don't have the ID collision problem at all. Either way, it's out of scope for this PR.